Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,634 @@
+  m_stacked_widget->addWidget(m_placeholder);
+  m_placeholder->setCursor(m_text_edit->cursor());
+  m_placeholder->setAttribute(Qt::WA_TransparentForMouseEvents);
+  m_scroll_box = new ScrollBox(m_stacked_widget, this);
+  m_scroll_box->set(ScrollBox::DisplayPolicy::NEVER,
+    ScrollBox::DisplayPolicy::ON_OVERFLOW);
+  m_scroll_box->get_vertical_scroll_bar().adjustSize();
+  m_scroll_box->setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
+  m_scroll_box->installEventFilter(this);
+  m_scroll_box->findChild<ScrollableLayer*>()->installEventFilter(this);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(m_scroll_box);
+  proxy_style(*this, *m_scroll_box);
+  add_pseudo_element(*this, Placeholder());
+  connect_style_signal(*this, [=] { on_style(); });
+  connect_style_signal(*this, Placeholder(), [=] { on_style(); });
+  set_style(*this, DEFAULT_STYLE());
+  connect(m_text_edit, &QTextEdit::cursorPositionChanged, this,
+    &TextAreaBox::on_cursor_position);
+  m_current_connection = m_model->connect_current_signal(
+    [=] (const auto& value) { on_current(value); });
+  connect(m_text_edit->document()->documentLayout(),
+    &QAbstractTextDocumentLayout::documentSizeChanged,
+    [=] (const auto& size) { m_stacked_widget->adjustSize(); });
+}
+
+const std::shared_ptr<TextModel>& TextAreaBox::get_model() const {
+  return m_model;
+}
+
+const QString& TextAreaBox::get_submission() const {
+  return m_submission;
+}
+
+void TextAreaBox::set_placeholder(const QString& value) {
+  m_placeholder_text = value;
+  update_placeholder_text();
+}
+
+bool TextAreaBox::is_read_only() const {
+  return m_text_edit->isReadOnly();
+}
+
+void TextAreaBox::set_read_only(bool read_only) {
+  m_text_edit->setReadOnly(read_only);
+  m_scroll_box->get_vertical_scroll_bar().set_position(0);
+  if(read_only) {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::NEVER);
+    match(*this, ReadOnly());
+  } else {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::ON_OVERFLOW);
+    unmatch(*this, ReadOnly());
+  }
+  update_layout();
+}
+
+connection TextAreaBox::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+QSize TextAreaBox::sizeHint() const {
+  return m_scroll_box->sizeHint();
+}
+
+void TextAreaBox::changeEvent(QEvent* event) {
+  if(event->type() == QEvent::EnabledChange) {
+    update_placeholder_text();
+  }
+  QWidget::changeEvent(event);
+}
+
+bool TextAreaBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_text_edit && event->type() == QEvent::FocusOut) {
+    m_submission = m_model->get_current();
+    m_submit_signal(m_submission);
+  } else if(event->type() == QEvent::Wheel) {
+    if(m_text_edit->isReadOnly()) {
+      return true;
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void TextAreaBox::mousePressEvent(QMouseEvent* event) {
+  m_text_edit->setFocus();
```

Comment:
Why is this needed? What happens if it's removed? Clicking on a widget should always give it focus by default.

Reply:
Even with the focus proxy set the text edit isn't receiving focus when the TextAreabox is clicked. I'll dig into it a little more.

Reply:
Sure, if there's a bug then create a task for it. Don't work around bugs without first reporting them, it results in everyone coming up with hacky ways of working around issues instead of solving it at the root.